### ISSUE TITLE
In Pascal-like languages begin-end brackets incorrectly match words containing them as parts

### ISSUE THREAD

**Issue by Somnium7** [microsoft/vscode#26121]
When defining Pascal-like language, with begin-end brackets, editor thinks, that "begin" and "end" being parts of any other words are also brackets.
I have tried to use regular expression for brackets, however, it seems that regex is not supported there.

![match](https://cloud.githubusercontent.com/assets/23244522/25773124/b42aef54-3280-11e7-8c39-d331cc745adf.png)

"language-configuration.json":
![lang-conf](https://cloud.githubusercontent.com/assets/23244522/25773129/c036890c-3280-11e7-9579-64250864713c.png)

- VSCode Version: 1.12.1
- OS Version: Windows 10

Steps to Reproduce:

1. Copy any language extension to custom extensions folder
2. In "language-configuration.json" add "begin-end" brackets
3. Write in a new file "begin-end" block and "appendElem" function call inside that block
4. Click on "begin" and see that it matches with function call

**Comment by zabel-xyz** [comment]
In PLSQL language we have delimiters inside others: begin..end but also if...end if

So I need to do something like this:
```
    "brackets": [
        ["if", "end if"],
        ["loop", "end loop"],
        ["begin", "end"]
    ],
```
=> It doesn't work with begin...end
![image](https://user-images.githubusercontent.com/16486407/32133630-7df8acba-bbdc-11e7-8773-d9e2ce4ce817.png)
![image](https://user-images.githubusercontent.com/16486407/32133643-d48e319e-bbdc-11e7-8283-54d83543ad8f.png)


I added character ;
```
    "brackets": [
        ["if", "end if;"],
        ["loop", "end loop;"],
        ["begin", "end;"]
    ],
```
=> it doesn't work anymore, nothing is higlighted this time.

**Comment by Somnium7** [comment]
Any progress on this one? Regex support will be ideal, however only checking word boundaries will be already great improvement.

**Comment by alexandrudima** [comment]
I've added checks for word boundaries via https://github.com/Microsoft/vscode/commit/91bd149bfe44b215882f843f6d0d925e40df0aab

**Comment by Somnium7** [comment]
@alexandrudima Thanks! Checked it in Insiders. Working almost great, one minor bug:
If we have two types of brackets which differ only by opening bracket, but closing bracket is same, then in some cases bracket are not higlighted when plasing cursor at closing brackets.

Take following configuration:
```
{
	"comments": {
		"lineComment": "//",
		"blockComment": ["/*", "*/"]
	},
	"brackets": [
		["(", ")"],
		["recordbegin", "endrecord"],
		["simplerecordbegin", "endrecord"]
	],
	"autoClosingPairs": [
		["(", ")"],
		["\"", "\""],
		["'", "'"]
	],
	"surroundingPairs": [
		["(", ")"],
		["\"", "\""],
		["'", "'"]
	]
}
```
And place cursor on each bracket to see what happens. See screeshot.

![bracketbug](https://user-images.githubusercontent.com/23244522/39145830-93510c50-473d-11e8-851c-42a7d6c6907a.png)

**Comment by Danieleeee** [comment]
Hi, still don't work in this case:
`"brackets": [
        ["if", "end if"],
        ["begin", "end"]
    ],`
![1](https://user-images.githubusercontent.com/22749175/39176953-6188171a-47ae-11e8-872c-46ed81e5911b.png)

**Comment by alexandrudima** [comment]
To verify, edit one of the built-in language configuration files or create a new language if you really want to and use:
```
 "brackets": [
        ["if", "end if"],
        ["loop", "end loop"],
        ["begin", "end"]
    ],
```

Then, write some code which utilizes this, like.
```
begin
  loop
    if
    end if
  end loop
end
```

The `end` should match the corresponding beginning sequence.

Please also try other variants from other comments...

**Comment by Somnium7** [comment]
@alexandrudima Thank you! I have tried in Insiders version some of examples which didn't work earlier and they all seems working now.

**Comment by Danieleeee** [comment]
![2](https://user-images.githubusercontent.com/22749175/67865576-c80b7080-fb27-11e9-9459-9096195ed8a1.jpg)

**Comment by Danieleeee** [comment]
Yes.
You have tried and it works ?

**Comment by alexandrudima** [comment]
@Danieleeee Yes, I have tried when I pushed fixes for this issue. There is even a unit test that passes for this case. See [here](https://github.com/microsoft/vscode/blob/f489e31c5bd3a36476544361d48c4b72fc3295ad/src/vs/editor/test/common/model/textModelWithTokens.test.ts#L260-L306)

I have tried once again with the brackets you have defined:
![image](https://user-images.githubusercontent.com/5047891/67938206-5424a300-fbcf-11e9-9a90-861ab0bd2649.png)

And it works for me:
![image](https://user-images.githubusercontent.com/5047891/67938251-6999cd00-fbcf-11e9-854c-c96e0a54bbd7.png)

Have you tried in a VS Code build that contains the fix? i.e. an Insiders build? The 1.40 version will ship to the stable channel sometime next week.
